### PR TITLE
Update virtual-dom typings with constructors and type-guards

### PR DESCRIPTION
**Pull request by styu**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/Matt-Esch/virtual-dom#example---creating-a-vtree-using-the-objects-directly, https://github.com/Matt-Esch/virtual-dom/tree/master/vnode
- [x] Increase the version number in the header if appropriate.

### CODE REVIEW

File: types/virtual-dom/index.d.ts [hunk 1]
```diff
@@ -42,13 +42,21 @@ declare namespace VirtualDOM {
     type: string; // 'VirtualNode'
   }
 
+  interface VNodeConstructor {
+    new (tagName: string, properties: VProperties, children: VTree[], key?: string, namespace?: string): VNode;
+  }
+
   interface VText {
     text: string;
     new(text: any): VText;
```
Comment (dphilipson) [inline types/virtual-dom/index.d.ts]: To my understanding, this existing line is wrong and should be removed (the `VTextContructor` being added here is the correct typing). This line makes it possible to write things like
```ts
const text = new VText("Hello");
const why = new text("World");
```
which is nonsense. This PR would be a good chance to fix up the old types.

The same could be done with the `VPatch` definition, but that's fine to leave for another PR.

Comment (styu) [reply types/virtual-dom/index.d.ts]: @dphilipson turns out `patch` doesn't have a new operator according to the library (it's just a function), so updated the typings accordingly

**Review by dphilipson — APPROVED** [review]

**Review by chbrown — APPROVED** [review]
LGTM & feel free to add your name to the bylines.

File: types/virtual-dom/index.d.ts (hunk 1 above)
Comment (styu) [reply types/virtual-dom/index.d.ts]: yeah I looked through the virtual-dom library and `patch` is used as a function and not as a constructor, so I just removed the `new` constructor.

**Review by kvcpr — APPROVED** [review]

**Pull request merged**